crates/sui-core/src/authority/authority_store.rs

```diff
@@ -1104,27 +1121,36 @@ impl<S: Eq + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
             .collect();
         info!(?sequenced, "locking");
 
-        let mut write_batch = self.sequenced.batch();
-        write_batch = write_batch.insert_batch(&self.sequenced, sequenced)?;
+        let mut write_batch = self.assigned_object_versions.batch();
+        write_batch = write_batch.insert_batch(&self.assigned_object_versions, sequenced)?;
         write_batch.write()?;
 
         Ok(())
     }
 
     /// Lock a sequence number for the shared objects of the input transaction. Also update the
     /// last consensus index.
+    /// This function must only be called from the consensus task (i.e. from handle_consensus_transaction).
     pub fn persist_certificate_and_lock_shared_objects(
         &self,
         certificate: CertifiedTransaction,
         consensus_index: ExecutionIndices,
     ) -> Result<(), SuiError> {
         // Make an iterator to save the certificate.
         let transaction_digest = *certificate.digest();
-        // let certificate_to_write = std::iter::once((transaction_digest, &certificate));
+
+        // Ensure that we only advance next_object_versions exactly once for every cert received from
+        // consensus.
+        if self
+            .consensus_message_processed
```

Comment:
Would we also advance the versions in the node sync execute path?

Reply:
no, the whole point is that we can't - it must advance always and only in response to consensus messages. That way, whenever shared locks are set due to a consensus message, it is guaranteed that the state of `next_object_versions` does not depend on whether we've executed a cert via the sync path previously.

Reply:
I see. Could you maybe add a comment somewhere to explain why it's safe/ok for `next_object_versions` to be somewhat inconsistent with `assigned_object_versions`? (since `assigned_object_versions` is set properly in all paths, while `next_object_versions` is only set from consensus path)

Reply:
Done